File: apps/ios/Voiid/Voiid/Networking/StoryService.swift
//
//  StoryService.swift
//  Voiid
//
//  Thin API layer for the Stories endpoints (backend/api/src/routes/stories.ts). Pure
//  transport + DTOs — it never touches crypto and never sees a media key. The per-device
//  ciphertexts it ships are built by ChatEngine's fan-out; the blob it presigns is already
//  ciphertext. See docs/STORIES_PROTOCOL.md.
//

import Foundation

@MainActor
final class StoryService {
    static let shared = StoryService()
    private let api = APIClient()
    private init() {}

    // MARK: - Wire DTOs

    /// One target device's key material in a fan-out (POST /stories and /:id/keys, and
    /// receipts). Matches the server's `{recipient_device_id, ciphertext(b64)}` shape.
    struct KeyEntry: Codable {
        let recipient_device_id: String
        let ciphertext: String
    }

    struct PresignUploadResp: Decodable { let key: String; let upload_url: String }

    struct PostStoryResp: Decodable {
        let story_id: String
        let created_at: String
        let expires_at: String
        var delivered_devices: Int = 0
    }

    /// A story row as the feed hands it back, plus THIS device's own key blob.
    struct FeedStory: Decodable {
        let story_id: String
        let author_id: String
        var author_device_id: String?
        let r2_key: String
        let media_mime: String
        var byte_size: Int?
        let created_at: String
        let expires_at: String
        let ciphertext: String        // Session.encrypt(envelope) for this device
    }
    struct FeedResp: Decodable { let stories: [FeedStory] }

    struct MineStory: Decodable {
        let story_id: String
        let r2_key: String
        let media_mime: String
        var byte_size: Int?
        let created_at: String
        let expires_at: String
        var recipient_device_count: Int = 0
        var delivered_device_count: Int = 0
    }
    struct MineResp: Decodable { let stories: [MineStory] }

    struct PresignDownloadResp: Decodable { let download_url: String }

    struct ReceiptRow: Decodable {
        let id: String
        let story_id: String
        let ciphertext: String
        let created_at: String
    }
    struct ReceiptsResp: Decodable { let receipts: [ReceiptRow] }

    // MARK: - Endpoints

    /// §7.1 — a presigned PUT for the story ciphertext. `mime` is the WRAPPER type; the
    /// real media type rides encrypted in the envelope.
    func presignUpload(mime: String = "application/octet-stream") async throws -> PresignUploadResp {
        struct Body: Encodable { let mime: String }
        return try await api.request("POST", "stories/presign-upload", body: Body(mime: mime))
    }

    /// §7.2 — create the story row + fan out per-device key blobs.
    func postStory(storyId: String, r2Key: String, mediaMime: String, byteSize: Int,
                   senderDeviceId: String?, keys: [KeyEntry]) async throws -> PostStoryResp {
        struct Body: Encodable {
            let story_id: String; let r2_key: String; let media_mime: String
            let byte_size: Int; let sender_device_id: String?; let keys: [KeyEntry]
        }
        return try await api.request("POST", "stories", body: Body(
            story_id: storyId, r2_key: r2Key, media_mime: mediaMime, byte_size: byteSize,
            sender_device_id: senderDeviceId, keys: keys))
    }

    /// §7.3 — this device's pending story key blobs (atomic mark-and-return).
    /// `includeDelivered` re-fetches already-delivered live rows after a reinstall.
    func feed(deviceId: String, includeDelivered: Bool = false) async throws -> [FeedStory] {
        var path = "stories/feed?device_id=\(deviceId)"
        if includeDelivered { path += "&include_delivered=1" }
        let resp: FeedResp = try await api.request("GET", path)
        return resp.stories
    }

    /// §7.4 — my own live stories + delivered-device counts.
    func mine(deviceId: String) async throws -> [MineStory] {
        let resp: MineResp = try await api.request("GET", "stories/mine?device_id=\(deviceId)")
        return resp.stories
    }

    /// §7.5 — a presigned GET for a story's ciphertext, entitlement-checked server-side.
    func presignDownload(storyId: String) async throws -> String {
        struct Body: Encodable { let story_id: String }
        let resp: PresignDownloadResp = try await api.request("POST", "stories/presign-download", body: Body(story_id: storyId))
        return resp.download_url
    }

    /// §7.6 — author-only delete. Not a security operation (already-downloaded copies survive).
    func delete(storyId: String) async throws {
        _ = try await api.request("DELETE", "stories/\(storyId)") as EmptyResponse
    }

    /// §7.7 — post a view receipt fanned out to the author's devices.
    func postReceipt(storyId: String, receipts: [KeyEntry]) async throws {
        struct Body: Encodable { let receipts: [KeyEntry] }
        _ = try await api.request("POST", "stories/\(storyId)/receipt", body: Body(receipts: receipts)) as EmptyResponse
    }

    /// §7.8 — append key material for devices skipped at post time / newly registered.
    func addKeys(storyId: String, keys: [KeyEntry]) async throws {
        struct Body: Encodable { let keys: [KeyEntry] }
        _ = try await api.request("POST", "stories/\(storyId)/keys", body: Body(keys: keys)) as EmptyResponse
    }

    /// §7.9 — an author device pulls pending view receipts (atomic mark-and-return).
    func receipts(deviceId: String) async throws -> [ReceiptRow] {
        let resp: ReceiptsResp = try await api.request("GET", "stories/receipts?device_id=\(deviceId)")
        return resp.receipts
    }
}

File: apps/android/app/src/main/java/com/voiid/app/net/StoryService.kt
package com.voiid.app.net

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Stories API transport (mirrors iOS StoryService). Every blob is encrypted ON-DEVICE before it
 * leaves; the server signs short-lived R2 URLs and stores CIPHERTEXT, opaque object keys, and one
 * opaque per-recipient-device key blob. It never sees the media key, the real mime, the caption,
 * the dimensions, or the bytes.
 *
 * The upload/download here deliberately do NOT reuse [MediaService]: stories have their own
 * presign endpoints (`stories/presign-upload`, `stories/presign-download`) so the download can be
 * entitlement-checked (author OR holds a story_keys row) instead of the generic media endpoint's
 * "authenticated AND key starts with media/" — defence in depth, not a guarantee.
 */
class StoryService(private val tokens: TokenStore) {
    private val api = ApiClient(tokens)

    // Longer timeouts for the direct R2 transfers (a story can be a 50 MB video ciphertext).
    private val blobClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .build()

    // MARK: - DTOs

    @Serializable private data class PresignUploadBody(val mime: String)
    @Serializable private data class PresignUploadResp(val key: String, val upload_url: String)

    @Serializable data class KeyEntry(val recipient_device_id: String, val ciphertext: String)
    @Serializable private data class PostStoryBody(
        val story_id: String,
        val r2_key: String,
        val media_mime: String,
        val byte_size: Long,
        val sender_device_id: String?,
        val keys: List<KeyEntry>,
    )
    @Serializable data class PostStoryResp(
        val story_id: String,
        val created_at: String? = null,
        val expires_at: String? = null,
        val delivered_devices: Int = 0,
    )

    @Serializable data class FeedStory(
        val story_id: String,
        val author_id: String,
        val author_device_id: String? = null,
        val r2_key: String,
        val media_mime: String,
        val byte_size: Long? = null,
        val created_at: String,
        val expires_at: String,
        val ciphertext: String,
    )
    @Serializable private data class FeedResp(val stories: List<FeedStory> = emptyList())

    @Serializable data class MineStory(
        val story_id: String,
        val r2_key: String,
        val media_mime: String,
        val byte_size: Long? = null,
        val created_at: String,
        val expires_at: String,
        val recipient_device_count: Int = 0,
        val delivered_device_count: Int = 0,
    )
    @Serializable private data class MineResp(val stories: List<MineStory> = emptyList())

    @Serializable private data class PresignDownloadBody(val story_id: String)
    @Serializable private data class PresignDownloadResp(val download_url: String)

    @Serializable private data class ReceiptBody(val receipts: List<KeyEntry>)
    @Serializable private data class KeysBody(val keys: List<KeyEntry>)

    @Serializable data class ReceiptRow(
        val id: String,
        val story_id: String,
        val ciphertext: String,
        val created_at: String? = null,
    )
    @Serializable private data class ReceiptsResp(val receipts: List<ReceiptRow> = emptyList())

    // MARK: - Calls

    /** Presign a PUT and push the CIPHERTEXT straight to R2. Returns the opaque object key. The
     *  mime handed to the presigner is the WRAPPER type; the real media type rides in the envelope. */
    suspend fun uploadCiphertext(ciphertext: ByteArray, wrapperMime: String = "application/octet-stream"): String {
        val body = ApiClient.json.encodeToString(PresignUploadBody.serializer(), PresignUploadBody(wrapperMime))
        val presign: PresignUploadResp = api.requestAs("POST", "stories/presign-upload", jsonBody = body)
        withContext(Dispatchers.IO) {
            val req = Request.Builder()
                .url(presign.upload_url)
                .put(ciphertext.toRequestBody(wrapperMime.toMediaType()))
                .build()
            blobClient.newCall(req).execute().use {
                if (!it.isSuccessful) throw ApiError.Http(it.code, "story upload failed (${it.code})")
            }
        }
        return presign.key
    }

    suspend fun postStory(
        storyId: String, r2Key: String, mediaMime: String, byteSize: Long,
        senderDeviceId: String?, keys: List<KeyEntry>,
    ): PostStoryResp {
        val body = ApiClient.json.encodeToString(
            PostStoryBody.serializer(),
            PostStoryBody(storyId, r2Key, mediaMime, byteSize, senderDeviceId, keys),
        )
        return api.requestAs("POST", "stories", jsonBody = body)
    }

    /** Append keys for devices skipped at post time (no prekey) or newly-registered inside 24h. */
    suspend fun appendKeys(storyId: String, keys: List<KeyEntry>): Int {
        if (keys.isEmpty()) return 0
        val body = ApiClient.json.encodeToString(KeysBody.serializer(), KeysBody(keys))
        val resp: AddedResp = api.requestAs("POST", "stories/$storyId/keys", jsonBody = body)
        return resp.added
    }
    @Serializable private data class AddedResp(val added: Int = 0)

    suspend fun feed(deviceId: String?, includeDelivered: Boolean = false): List<FeedStory> {
        val q = buildString {
            append("stories/feed")
            val params = buildList {
                deviceId?.let { add("device_id=$it") }
                if (includeDelivered) add("include_delivered=1")
            }
            if (params.isNotEmpty()) append("?" + params.joinToString("&"))
        }
        val resp: FeedResp = api.requestAs("GET", q)
        return resp.stories
    }

    suspend fun mine(deviceId: String?): List<MineStory> {
        val q = "stories/mine" + (deviceId?.let { "?device_id=$it" } ?: "")
        val resp: MineResp = api.requestAs("GET", q)
        return resp.stories
    }

    /** Presign a download of a story's ciphertext and fetch it. Throws ApiError.Http on 403/404. */
    suspend fun downloadCiphertext(storyId: String): ByteArray {
        val body = ApiClient.json.encodeToString(PresignDownloadBody.serializer(), PresignDownloadBody(storyId))
        val presign: PresignDownloadResp = api.requestAs("POST", "stories/presign-download", jsonBody = body)
        return withContext(Dispatchers.IO) {
            val req = Request.Builder().url(presign.download_url).get().build()
            blobClient.newCall(req).execute().use {
                if (!it.isSuccessful) throw ApiError.Http(it.code, "story download failed (${it.code})")
                it.body?.bytes() ?: ByteArray(0)
            }
        }
    }

    suspend fun deleteStory(storyId: String) {
        api.request("DELETE", "stories/$storyId")
    }

    suspend fun sendReceipt(storyId: String, receipts: List<KeyEntry>): Int {
        if (receipts.isEmpty()) return 0
        val body = ApiClient.json.encodeToString(ReceiptBody.serializer(), ReceiptBody(receipts))
        val resp: AcceptedResp = api.requestAs("POST", "stories/$storyId/receipt", jsonBody = body)
        return resp.accepted
    }
    @Serializable private data class AcceptedResp(val accepted: Int = 0)

    suspend fun receipts(deviceId: String?): List<ReceiptRow> {
        val q = "stories/receipts" + (deviceId?.let { "?device_id=$it" } ?: "")
        val resp: ReceiptsResp = api.requestAs("GET", q)
        return resp.receipts
    }
}
